### PR TITLE
[WIP] Allow reordering of pages by dragging the carousel indicators

### LINKED ISSUE #2525: Allow reordering of pages by dragging the carousel indicators
Underneath the pages there are little bars for every page in the story.

For faster creation of stories it could be useful to be able to quickly reorder pages by dragging around these editor carousel indicators.

### PR DESCRIPTION
**Pull request by ravinderk**
Fixes #2525

Todo:
- [x] Create `ReordererIndicator` component
- [x] Add action to set `order` in `amp/story` before render `ReordererIndicator`
- [x] Remove deprecated `Indicator` component
- [ ] E2E tests.

### CODE REVIEW

**Comment by ravinderk** [comment]
I signed it!

File: assets/src/stories-editor/components/editor-carousel/reordererIndicator.js [hunk 1]
```diff
@@ -0,0 +1,185 @@
+/**
```
Comment (swissspidy) [inline assets/src/stories-editor/components/editor-carousel/reordererIndicator.js]: Why was this renamed from `indicator.js` to `reordererIndicator.js`?

We do not really use such naming for other components and their files. If the component is named `ReordererIndicator`, then `reorderer-indiccator.js` would be more fitting.

However, I don't see why it was renamed. Sure, the component now has a new job, but this name change seems misleading and unneeded IMO.

File: assets/src/stories-editor/components/editor-carousel/reordererIndicator.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+			return;
+		}
+
+		const positionIndex = this.getInsertIndex( position );
+		const insertIndex = srcIndex < index ? positionIndex - 1 : positionIndex;
+		movePageToPosition( srcClientId, insertIndex );
+	}
+
+	componentDidMount() {
+		this.props.setOrder();
```
Comment (swissspidy) [inline assets/src/stories-editor/components/editor-carousel/reordererIndicator.js]: The whole `setOrder` logic here seems... confusing.

I think we should handle this differently in a more simplified way: 

* `movePageToPosition` should just save the current block order in the block editor store permanently, not using our own store at all.
* No need for `setOrder`

In the regular reorder mode we could also think about simplifying the logic. Maybe using our own data store is a bit unnecessary.

File: assets/src/stories-editor/components/editor-carousel/reordererIndicator.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+		/* translators: %s: Page number */
+		const label = ( pageNumber ) => sprintf( __( 'Page %s', 'amp' ), pageNumber );
+
+		/* translators: %s: Page number */
+		const toolTip = ( pageNumber ) => sprintf( __( 'Go to page %s', 'amp' ), pageNumber );
+		const className = page.clientId === currentPage ? 'amp-story-editor-carousel-item amp-story-editor-carousel-item--active' : 'amp-story-editor-carousel-item';
+
+		return (
+			<div className="amp-story-reorderer-carousel-item">
```
Comment (swissspidy) [inline assets/src/stories-editor/components/editor-carousel/reordererIndicator.js]: What's the reason for not using `<ul>` and `<li>` anymore? Was something not working?

File: assets/src/stories-editor/components/editor-carousel/reordererIndicator.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+			srcIndex: index,
+			srcClientId: clientId,
+		};
+
+		/* translators: %s: Page number */
+		const label = ( pageNumber ) => sprintf( __( 'Page %s', 'amp' ), pageNumber );
+
+		/* translators: %s: Page number */
+		const toolTip = ( pageNumber ) => sprintf( __( 'Go to page %s', 'amp' ), pageNumber );
+		const className = page.clientId === currentPage ? 'amp-story-editor-carousel-item amp-story-editor-carousel-item--active' : 'amp-story-editor-carousel-item';
```
Comment (swissspidy) [inline assets/src/stories-editor/components/editor-carousel/reordererIndicator.js]: Let's use the `classnames` function here while we're at it

**Comment by swissspidy** [comment]
Thanks for working on this! Left a few comments while glancing over the code. Maybe we can discuss the data store part a bit more in-depth.

File: assets/src/stories-editor/components/editor-carousel/reordererIndicator.js (hunk 1 above)
Comment (ravinderk) [reply assets/src/stories-editor/components/editor-carousel/reordererIndicator.js]: @swissspidy Agree on this, I will rename it back to `indicator.js`

File: assets/src/stories-editor/components/editor-carousel/reordererIndicator.js (hunk 4 above)
Comment (ravinderk) [reply assets/src/stories-editor/components/editor-carousel/reordererIndicator.js]: 👍

File: assets/src/stories-editor/components/editor-carousel/reordererIndicator.js (hunk 3 above)
Comment (ravinderk) [reply assets/src/stories-editor/components/editor-carousel/reordererIndicator.js]: @swissspidy `DropZone` component by default generates `div` and we are outputting that just before every indicator item in `indicator.js`. `li` and `div` within `ul` is not a good DOM structure, so move to use only `div`.

File: assets/src/stories-editor/components/editor-carousel/reordererIndicator.js (hunk 2 above)
Comment (ravinderk) [reply assets/src/stories-editor/components/editor-carousel/reordererIndicator.js]: @swissspidy I think, `resetOrder` is useful in the current scenario 👍

**Comment by ravinderk** [comment]
@swissspidy thank you for detail review. I completed most of the review. Please review my comments and let me know if you have any suggestion or question.

File: assets/src/stories-editor/components/editor-carousel/reordererIndicator.js (hunk 3 above)
Comment (swissspidy) [reply assets/src/stories-editor/components/editor-carousel/reordererIndicator.js]: What is bad about

```html
<ul>
	<li>
		<div> ... </div>
	</li>
</ul>
```

That is more semantic than just using divs everywhere

File: assets/src/stories-editor/components/editor-carousel/reordererIndicator.js (hunk 2 above)
Comment (swissspidy) [reply assets/src/stories-editor/components/editor-carousel/reordererIndicator.js]: That's not what I meant.

I don't think this component should have a side effect of calling `setOrder` or `resetOrder` inside `componentDidMount`. 

I would remove this, and then instead of `movePageToPosition`, just use `moveBlockToPosition` from `core/block-editor`.

**Comment by miina** [comment]
Additionally, did some functional testing, some findings:
- Currently, the dropzone seems to be matching the indicator itself, it's not clear where the Page is being moved, to which side. It would make more sense if the dropzone would display more in between the indicators or on the sides (see how it is when using the Reordering for example).
- It doesn't seem to be possible to move the currently selected Page.
- To consider: it would be good to have a Preview of the Page when hovering over a Page indicator -- it's quite confusing right now to understand where was the correct place for copying, especially when having multiple pages.
- Once all is working as expected then it might be worthwhile to check if there are parts that we could reuse/share between the Reorderer and the Indicator component -- the parts that are copied 1-1.

File: assets/src/stories-editor/components/editor-carousel/indicator.js [hunk 5]
```diff
@@ -58,11 +58,12 @@ class Indicator extends Component {
...
 
 	getInsertIndex( position ) {
-		const { index } = this.props;
+		let { index } = this.props;
 
 		if ( index !== undefined ) {
-			return position.x === 'right' ? index + 1 : index;
+			index = position.x === 'right' ? index + 1 : index;
 		}
+		return index;
```
Comment (miina) [inline assets/src/stories-editor/components/editor-carousel/indicator.js]: Could you clarify why was this change necessary?

Comment (swissspidy) [reply assets/src/stories-editor/components/editor-carousel/indicator.js]: The new ESLint rules I added require always having an explicit return.

Although here I'd just `return undefined` instead of `return index`, even though it's technically the same. But it's eaiser to scan.

Comment (miina) [reply assets/src/stories-editor/components/editor-carousel/indicator.js]: Yep, we could just use return here instead of the other changes above.

File: assets/src/stories-editor/components/editor-carousel/reordererIndicator.js (hunk 3 above)
Comment (ravinderk) [reply assets/src/stories-editor/components/editor-carousel/reordererIndicator.js]: @swissspidy You are right. I thought that `li` can not hold `div` and bump out it. I revert back to the same DOM (`ul-li`) structure with drag and drop functionality.

image

ref: https://stackoverflow.com/questions/6449772/can-i-use-a-div-inside-a-list-item

File: assets/src/stories-editor/components/editor-carousel/indicator.js (hunk 5 above)
Comment (ravinderk) [reply assets/src/stories-editor/components/editor-carousel/indicator.js]: code updated

**Comment by ravinderk** [comment]
@miina 

 - **Currently, the dropzone seems to be matching the indicator itself, it's not clear where the Page is being moved, to which side. It would make more sense if the dropzone would display more in between the indicators or on the sides (see how it is when using the Reordering for example).**
This has been fixed. But I am having weird styling issue with the indicator when you drag it.
- **It doesn't seem to be possible to move the currently selected Page.**
I removed the `disabled` tag from button to make active indicator draggable.
- **To consider: it would be good to have a Preview of the Page when hovering over a Page indicator -- it's quite confusing right now to understand where was the correct place for copying, especially when having multiple pages.**
Agree on this. It is a useful feature but needs to think about the tooltip also. Like how we will show tooltip and page information on hover together. 
- **Once all is working as expected then it might be worthwhile to check if there are parts that we could reuse/share between the Reorderer and the Indicator component -- the parts that are copied 1-1.** 
Agree on this 👍

**Comment by ravinderk** [comment]
@miina 
- **Currently, the dropzone seems to be matching the indicator itself, it's not clear where the Page is being moved, to which side. It would make more sense if the dropzone would display more in between the indicators or on the sides (see how it is when using the Reordering for example).**
this has been fixed but I am facing weird styling issue of indicator when I drag it.
- **It doesn't seem to be possible to move the currently selected Page.**
I removed the `disabled` tag from button to make is draggable. 
- **To consider: it would be good to have a Preview of the Page when hovering over a Page indicator -- it's quite confusing right now to understand where was the correct place for copying, especially when having multiple pages.**
Agree on this but need to decide how this will work because currently, we are showing a tooltip.
- **Once all is working as expected then it might be worthwhile to check if there are parts that we could reuse/share between the Reorderer and the Indicator component -- the parts that are copied 1-1.**
Agree on this.

**Comment by swissspidy** [comment]
> it would be good to have a Preview of the Page when hovering over a Page indicator

There's a separate issue for this, see #2526.

**Comment by swissspidy** [comment]
Closing this for now. Can be picked up when #2525 gains traction again.
